File: wiki-plugin/scripts/tool_call_stats.py
"""Summarise the tool-call log written by hooks/log_tool_calls.py (#100).

Makes the cost of a run visible: total tool calls, a per-tool histogram,
and (as a proxy — see below) how many user-prompt turns those calls fell
into and the calls-per-turn ratio.

Per #99's spike, the PostToolUse payload carries no per-assistant-message
identifier and no timestamp, so exact assistant-turn count is not
recoverable from the log. ``prompt_id`` is the closest available grouping
key, but it's scoped to a whole user-prompt turn (which may itself span
several assistant turns), so "prompts" here is a coarser, honestly-labelled
proxy — not an exact turn count.

CLI::

    python tool_call_stats.py [--session-id <id>]

Defaults ``--session-id`` to ``$CLAUDE_CODE_SESSION_ID``, which Claude Code
exports to every Bash tool call it makes.
"""
from __future__ import annotations

import argparse
import json
import os
import sys
from collections import Counter
from pathlib import Path
from typing import Sequence

from session_state import sessions_dir


def log_path(session_id: str, state_dir: Path | None = None) -> Path:
    return (state_dir if state_dir is not None else sessions_dir()) / f"{session_id}-tool-calls.jsonl"


def read_log(session_id: str, state_dir: Path | None = None) -> list[dict]:
    """The logged events for ``session_id``, oldest first. ``[]`` if no log exists."""
    path = log_path(session_id, state_dir)
    if not path.is_file():
        return []
    events = []
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            events.append(json.loads(line))
        except json.JSONDecodeError:
            continue
    return events


def summarize(events: list[dict]) -> dict:
    """Aggregate ``events`` into totals, a per-tool histogram, and the prompt-count proxy."""
    total = len(events)
    by_tool = Counter(e.get("tool") or "?" for e in events)
    prompt_ids = {e.get("prompt_id") for e in events if e.get("prompt_id")}
    prompts = len(prompt_ids)
    return {
        "total": total,
        "by_tool": by_tool.most_common(),
        "prompts": prompts,
        "calls_per_prompt": (total / prompts) if prompts else None,
    }


def format_summary(stats: dict) -> str:
    lines = [f"Total tool calls: {stats['total']}"]
    for tool, n in stats["by_tool"]:
        lines.append(f"  {n:3}  {tool}")
    if stats["prompts"]:
        lines.append(
            f"Prompts (proxy for turns, not exact — see #99): {stats['prompts']}, "
            f"{stats['calls_per_prompt']:.1f} calls/prompt"
        )
    return "\n".join(lines)


def _main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--session-id", default=os.environ.get("CLAUDE_CODE_SESSION_ID"),
        help="session to summarise (default: $CLAUDE_CODE_SESSION_ID)",
    )
    args = parser.parse_args(argv)
    if not args.session_id:
        sys.exit("No session_id — pass --session-id or set $CLAUDE_CODE_SESSION_ID")

    events = read_log(args.session_id)
    if not events:
        sys.exit(f"No log found at {log_path(args.session_id)}")

    print(format_summary(summarize(events)))
    return 0


if __name__ == "__main__":  # pragma: no cover
    sys.exit(_main())

File: enchiridion-go/internal/toolcallstats/toolcallstats.go
// Package toolcallstats summarises the tool-call log written by the
// PostToolUse hook. Ported from `wiki-plugin/scripts/tool_call_stats.py`.
//
// Makes a run's cost visible: total tool calls, a per-tool histogram, and a
// prompt count with calls-per-prompt.
//
// **"Prompts" is a proxy, not a turn count** (#99). The PostToolUse payload
// carries no per-assistant-message identifier and no timestamp, so exact
// assistant turns aren't recoverable. prompt_id is the closest grouping key
// available, but it spans a whole user-prompt turn — which may itself cover
// several assistant turns. Labelled honestly wherever it's printed.
package toolcallstats

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/dhague/enchiridion/enchiridion-go/internal/sessionstate"
)

// LogPath returns the tool-call log path for sessionID under stateDir. An
// empty stateDir resolves the session state directory, matching the Python
// `state_dir or sessions_dir()` default.
func LogPath(sessionID, stateDir string) string {
	if stateDir == "" {
		stateDir = sessionstate.SessionsDir("", "", nil)
	}
	return filepath.Join(stateDir, sessionID+"-tool-calls.jsonl")
}

// ReadLog returns the logged events for sessionID, oldest first. An empty
// slice when no log exists. Malformed or blank lines are skipped.
func ReadLog(sessionID, stateDir string) ([]map[string]any, error) {
	data, err := os.ReadFile(LogPath(sessionID, stateDir))
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var events []map[string]any
	for _, line := range strings.Split(string(data), "\n") {
		if strings.TrimSpace(line) == "" {
			continue
		}
		var event map[string]any
		if err := json.Unmarshal([]byte(line), &event); err != nil {
			continue
		}
		events = append(events, event)
	}
	return events, nil
}

// ToolCount is one tool with the number of calls, in histogram order.
type ToolCount struct {
	Tool  string
	Count int
}

// Summary is the aggregate of one run's tool-call log.
type Summary struct {
	Total int
	// ByTool is the per-tool histogram, most-called first (ties broken by
	// first-seen order, matching Python's Counter.most_common).
	ByTool []ToolCount
	// Prompts is the prompt-count proxy; see the package comment.
	Prompts int
	// CallsPerPrompt is valid only when HasCallsPerPrompt is true (i.e. there
	// is at least one prompt to divide by).
	CallsPerPrompt    float64
	HasCallsPerPrompt bool
}

// Summarize aggregates events into totals, a per-tool histogram, and the
// prompt-count proxy.
func Summarize(events []map[string]any) Summary {
	total := len(events)

	counts := map[string]int{}
	order := []string{}
	promptIDs := map[string]bool{}
	for _, event := range events {
		tool, _ := event["tool"].(string)
		if tool == "" {
			tool = "?"
		}
		if _, seen := counts[tool]; !seen {
			order = append(order, tool)
		}
		counts[tool]++
		if id, ok := event["prompt_id"].(string); ok && id != "" {
			promptIDs[id] = true
		}
	}

	byTool := make([]ToolCount, 0, len(order))
	for _, tool := range order {
		byTool = append(byTool, ToolCount{Tool: tool, Count: counts[tool]})
	}
	// Stable sort by count descending, preserving first-seen order on ties.
	for i := 1; i < len(byTool); i++ {
		for j := i; j > 0 && byTool[j].Count > byTool[j-1].Count; j-- {
			byTool[j], byTool[j-1] = byTool[j-1], byTool[j]
		}
	}

	prompts := len(promptIDs)
	s := Summary{Total: total, ByTool: byTool, Prompts: prompts}
	if prompts > 0 {
		s.CallsPerPrompt = float64(total) / float64(prompts)
		s.HasCallsPerPrompt = true
	}
	return s
}

// FormatSummary renders a summary as the fixed text the CLI prints.
func FormatSummary(s Summary) string {
	var b strings.Builder
	fmt.Fprintf(&b, "Total tool calls: %d\n", s.Total)
	for _, tc := range s.ByTool {
		fmt.Fprintf(&b, "  %3d  %s\n", tc.Count, tc.Tool)
	}
	if s.Prompts > 0 {
		fmt.Fprintf(&b,
			"Prompts (proxy for turns, not exact — see #99): %d, %.1f calls/prompt",
			s.Prompts, s.CallsPerPrompt)
	}
	return b.String()
}
